From 6c5472237c3b095773a6012ed9483d01a8f91816 Mon Sep 17 00:00:00 2001
From: Artem Shilkin <89970996+reshilkin@users.noreply.github.com>
Date: Tue, 29 Aug 2023 20:28:26 +0300
Subject: [PATCH 4/4] Fix compilation with newer flatbuffers (#17164)

In flatbuffers@v23.5.9 was broken forward declaration for
FlatBufferBuilder. Trying to compile onnxruntime falls with the
following error:
```
flatbuffers/include/flatbuffers/flatbuffer_builder.h:1420:38: error: typedef redefinition with different types ('FlatBufferBuilderImpl<false>' vs 'flatbuffers::FlatBufferBuilder')
typedef FlatBufferBuilderImpl<false> FlatBufferBuilder;
                                     ^
onnx_runtime/include/onnxruntime/core/graph/graph.h:47:11: note: previous definition is here
    class FlatBufferBuilder;
```
This PR removes these declarations and puts includes instead

Upstream-Status: Inappropriate [Yocto specific]
---
 cmake/onnxruntime_providers.cmake                  |  2 +-
 include/onnxruntime/core/graph/graph.h             |  8 ++------
 onnxruntime/core/flatbuffers/flatbuffers_utils.h   | 14 ++------------
 .../core/framework/kernel_type_str_resolver.h      |  8 ++------
 onnxruntime/core/framework/session_state.h         |  8 ++------
 onnxruntime/core/graph/graph_flatbuffers_utils.h   |  8 ++------
 onnxruntime/core/graph/model.h                     |  9 +++------
 onnxruntime/core/graph/op_identifier_utils.h       | 11 ++---------
 .../graph/runtime_optimization_record_container.h  | 10 ++--------
 9 files changed, 18 insertions(+), 60 deletions(-)

diff --git a/cmake/onnxruntime_providers.cmake b/cmake/onnxruntime_providers.cmake
index 03360ff30c..e285e6b8b8 100644
--- a/cmake/onnxruntime_providers.cmake
+++ b/cmake/onnxruntime_providers.cmake
@@ -1799,7 +1799,7 @@ if (onnxruntime_USE_XNNPACK)
   source_group(TREE ${REPO_ROOT} FILES ${onnxruntime_providers_xnnpack_cc_srcs})
   onnxruntime_add_static_library(onnxruntime_providers_xnnpack ${onnxruntime_providers_xnnpack_cc_srcs})
   onnxruntime_add_include_to_target(onnxruntime_providers_xnnpack
-    onnxruntime_common onnxruntime_framework onnx onnx_proto ${PROTOBUF_LIB} XNNPACK pthreadpool Boost::mp11 safeint_interface
+    onnxruntime_common onnxruntime_framework onnx onnx_proto ${PROTOBUF_LIB} XNNPACK pthreadpool flatbuffers::flatbuffers Boost::mp11 safeint_interface
   )
 
   add_dependencies(onnxruntime_providers_xnnpack onnx ${onnxruntime_EXTERNAL_DEPENDENCIES})
diff --git a/include/onnxruntime/core/graph/graph.h b/include/onnxruntime/core/graph/graph.h
index 81015b25bc..19caa69d94 100644
--- a/include/onnxruntime/core/graph/graph.h
+++ b/include/onnxruntime/core/graph/graph.h
@@ -20,6 +20,8 @@
 #pragma warning(pop)
 #endif
 
+#include "flatbuffers/flatbuffers.h"
+
 #include "core/common/gsl.h"
 
 #include "core/common/common.h"
@@ -43,12 +45,6 @@
 #include "core/graph/node_arg.h"
 #include "core/graph/ort_format_load_options.h"
 
-namespace flatbuffers {
-class FlatBufferBuilder;
-template <typename T>
-struct Offset;
-}  // namespace flatbuffers
-
 namespace onnxruntime {
 class Graph;
 struct IndexedSubGraph;
diff --git a/onnxruntime/core/flatbuffers/flatbuffers_utils.h b/onnxruntime/core/flatbuffers/flatbuffers_utils.h
index 4e7db4df9a..55bde0b2df 100644
--- a/onnxruntime/core/flatbuffers/flatbuffers_utils.h
+++ b/onnxruntime/core/flatbuffers/flatbuffers_utils.h
@@ -5,6 +5,8 @@
 
 #include <unordered_map>
 
+#include "flatbuffers/flatbuffers.h"
+
 #include "core/common/common.h"
 #include "core/common/path_string.h"
 #include "core/common/status.h"
@@ -13,18 +15,6 @@ namespace ONNX_NAMESPACE {
 class ValueInfoProto;
 }
 
-namespace flatbuffers {
-class FlatBufferBuilder;
-
-template <typename T>
-struct Offset;
-
-struct String;
-
-template <typename T>
-class Vector;
-}  // namespace flatbuffers
-
 namespace onnxruntime {
 
 namespace fbs {
diff --git a/onnxruntime/core/framework/kernel_type_str_resolver.h b/onnxruntime/core/framework/kernel_type_str_resolver.h
index 75fc2fa894..31a806dd52 100644
--- a/onnxruntime/core/framework/kernel_type_str_resolver.h
+++ b/onnxruntime/core/framework/kernel_type_str_resolver.h
@@ -7,6 +7,8 @@
 #include <string_view>
 #include <utility>
 
+#include "flatbuffers/flatbuffers.h"
+
 #if !defined(ORT_MINIMAL_BUILD)
 #include "core/graph/onnx_protobuf.h"
 #endif  // !defined(ORT_MINIMAL_BUILD)
@@ -18,12 +20,6 @@
 #include "core/graph/graph.h"
 #include "core/platform/ort_mutex.h"
 
-namespace flatbuffers {
-class FlatBufferBuilder;
-template <typename T>
-struct Offset;
-}  // namespace flatbuffers
-
 namespace onnxruntime {
 
 namespace fbs {
diff --git a/onnxruntime/core/framework/session_state.h b/onnxruntime/core/framework/session_state.h
index d546f264a9..51bb02918d 100644
--- a/onnxruntime/core/framework/session_state.h
+++ b/onnxruntime/core/framework/session_state.h
@@ -8,6 +8,8 @@
 #include <unordered_map>
 #include <vector>
 
+#include "flatbuffers/flatbuffers.h"
+
 #include "core/common/gsl.h"
 
 #include "core/common/common.h"
@@ -43,12 +45,6 @@
 #include "core/framework/program_region.h"
 #endif
 
-namespace flatbuffers {
-class FlatBufferBuilder;
-template <typename T>
-struct Offset;
-}  // namespace flatbuffers
-
 namespace onnxruntime {
 
 namespace fbs {
diff --git a/onnxruntime/core/graph/graph_flatbuffers_utils.h b/onnxruntime/core/graph/graph_flatbuffers_utils.h
index f4899ffc12..b625cbf3ca 100644
--- a/onnxruntime/core/graph/graph_flatbuffers_utils.h
+++ b/onnxruntime/core/graph/graph_flatbuffers_utils.h
@@ -5,6 +5,8 @@
 
 #include <memory>
 
+#include "flatbuffers/flatbuffers.h"
+
 #include "core/common/status.h"
 #include "core/graph/ort_format_load_options.h"
 #include "core/framework/tensor.h"
@@ -18,12 +20,6 @@ class SparseTensorProto;
 #endif  // !defined(DISABLE_SPARSE_TENSORS)
 }  // namespace ONNX_NAMESPACE
 
-namespace flatbuffers {
-class FlatBufferBuilder;
-template <typename T>
-struct Offset;
-}  // namespace flatbuffers
-
 namespace onnxruntime {
 
 class Graph;
diff --git a/onnxruntime/core/graph/model.h b/onnxruntime/core/graph/model.h
index 5337211ae7..7e3942b029 100644
--- a/onnxruntime/core/graph/model.h
+++ b/onnxruntime/core/graph/model.h
@@ -7,6 +7,9 @@
 #include <memory>
 #include <climits>
 #include <string>
+
+#include "flatbuffers/flatbuffers.h"
+
 #include "core/common/path.h"
 #include "core/graph/graph_viewer.h"
 #include "core/graph/ort_format_load_options.h"
@@ -15,12 +18,6 @@
 #include "core/graph/function_template.h"
 #endif
 
-namespace flatbuffers {
-class FlatBufferBuilder;
-template <typename T>
-struct Offset;
-}  // namespace flatbuffers
-
 namespace onnxruntime {
 
 namespace fbs {
diff --git a/onnxruntime/core/graph/op_identifier_utils.h b/onnxruntime/core/graph/op_identifier_utils.h
index 265364a88d..8a9351a2d0 100644
--- a/onnxruntime/core/graph/op_identifier_utils.h
+++ b/onnxruntime/core/graph/op_identifier_utils.h
@@ -3,21 +3,14 @@
 
 #pragma once
 
+#include "flatbuffers/flatbuffers.h"
+
 #include "core/graph/op_identifier.h"
 
 #include "core/common/status.h"
 #include "core/graph/graph.h"
 #include "core/graph/onnx_protobuf.h"
 
-namespace flatbuffers {
-class FlatBufferBuilder;
-
-template <typename T>
-struct Offset;
-
-struct String;
-}  // namespace flatbuffers
-
 namespace onnxruntime {
 
 namespace fbs::utils {
diff --git a/onnxruntime/core/graph/runtime_optimization_record_container.h b/onnxruntime/core/graph/runtime_optimization_record_container.h
index 5db784f1a2..a28b19e786 100644
--- a/onnxruntime/core/graph/runtime_optimization_record_container.h
+++ b/onnxruntime/core/graph/runtime_optimization_record_container.h
@@ -9,17 +9,11 @@
 #include <unordered_map>
 #include <vector>
 
+#include "flatbuffers/flatbuffers.h"
+
 #include "core/common/common.h"
 #include "core/graph/runtime_optimization_record.h"
 
-namespace flatbuffers {
-class FlatBufferBuilder;
-template <typename T>
-struct Offset;
-template <typename T>
-class Vector;
-}  // namespace flatbuffers
-
 namespace onnxruntime {
 
 namespace fbs {
-- 
2.25.1

